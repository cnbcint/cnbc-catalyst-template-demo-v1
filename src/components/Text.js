import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";

const Pdrop = styled.p`
  color: ${(props) => props.color};
  font-size: ${(props) => (props.highlight ? "1.4em" : "1.2rem")};
  line-height: ${(props) => (props.highlight ? "140%" : "125%")};
  text-align: ${(props) => props.textAlign};
  max-width: 600px;
  margin:0 auto;
  &::first-letter {
    color: ${(props) => props.dropCapColor};
    padding: 0 0;
    margin: 0.5rem 1rem 0 0;
    font-size: 5.5rem;
    font-weight: bold;
    float: left;
    line-height: 4rem;
  }
`;
const P = styled.p`
  color: ${(props) => props.color};
  font-size: ${(props) => (props.highlight ? "1.4rem" : "1.2rem")};
  line-height: ${(props) => (props.highlight ? "140%" : "125%")};
  text-align: ${(props) => props.textAlign};
  max-width: 600px;
  margin:0 auto;

`;
/** 
 Text Block will add a paragraph element to it's children. Do not add P inside this element to avoid double nested P element
*/
export default function Text(props) {
  if (props.dropCap) {
    return (
      <Pdrop
        highlight={props.highlight}
        dropCapColor={props.dropCapColor ? props.dropCapColor : props.color}
        color={props.color}
        textAlign={props.textAlign}
        style={props.style}
        className={props.className}
      >
        {props.children}
      </Pdrop>
    );
  }
  return (
    <P
      highlight={props.highlight}
      color={props.color}
      style={props.style}
      textAlign={props.textAlign}
      className={props.className}
    >
      {props.children}
    </P>
  );
}

Text.defaulProps = {
  dropCap: false,
  color: "inherit",
  dropCapColor: "",
  highlight: false,
  className: "",
  textAlign: "left",
};

Text.propTypes = {
  /**
   Will Increase font size to highlight the text
   */
  highlight: PropTypes.bool,
  /**
   Will Increase font size of the first letter of the paragraph to achive DropCap 
   */
  dropCap: PropTypes.bool,
  /**
   CSS override for font color
   */
  color: PropTypes.string,
  /**
   CSS text aligment
   */
  textAlign: PropTypes.oneOf(["left", "right", "center"]),
  /**
   CSS color for drop cap
   */
  dropCapColor: PropTypes.string,
  /**
  React CSS format to override rules has required by design
  */
  style: PropTypes.object,
  className: PropTypes.string,
  children: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.array,
    PropTypes.element,
  ]).isRequired,
};
