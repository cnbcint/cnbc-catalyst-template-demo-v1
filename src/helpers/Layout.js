export function getJustifyContent(aligment) {
  switch (aligment) {
    case "left":
      return "flex-start";
    case "right":
      return "flex-end";
    default:
      return "center";
  }
}

export function getAnimation(aligment) {
  switch (aligment) {
    case "left":
      return "fadeInLeft";
    case "right":
      return "fadeInRight";
    default:
      return "fadeInDown";
  }
}

export const breakpoints = {
  mobile: "768px",
  tablet: "1280px",
};

export const margin = {
  base: 1,
  unit: "rem",
};

export function getMargin(multiplyer) {
  return multiplyer * margin.base + margin.unit;
}
