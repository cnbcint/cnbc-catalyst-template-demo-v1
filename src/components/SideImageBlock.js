import React from "react";
import styled from "styled-components";
import PropTypes from "prop-types";
import { breakpoints, getMargin } from "../helpers/Layout";

const Container = styled.div`
  width: 100%;
  display: flex;
  flex-direction: column;

  @media (min-width: ${breakpoints.mobile}) {
    flex-direction: ${(props) =>
      props.imageBlockPosition === "left" ? "row" : "row-reverse"};
    align-items: stretch;
  }
  justify-content: center;
  position: relative;
  box-sizing: border-box;

  margin-top: ${getMargin(3)};
  margin-bottom: ${getMargin(3)};
  @media (min-width: ${breakpoints.mobile}) {
    margin-top: ${getMargin(5)};
    margin-bottom: ${getMargin(5)};
  }
`;
const ContentBlock = styled.div`
  width: 100%;
  padding: ${getMargin(1)};
  box-sizing: border-box;
  display: flex;
  justify-content: center;
  align-items: center;
  background-color: ${(props) => props.backgroundColor};
  @media (min-width: ${breakpoints.mobile}) {
    padding: ${getMargin(3)};
    width: ${(props) => props.imageBlockWidth};
  }
`;
const ContentContainer = styled.div`
  @media (min-width: ${breakpoints.mobile}) {
    max-width: ${(props) => props.contentMaxWidth};
  }
`;

const ImageBlock = styled.div`
  background: black;
  background-image: url(${(props) => props.image.mobile});
  background-position: center;
  background-size: cover;
  img {
    margin: 0;
  }

  @media (min-width: ${breakpoints.mobile}) {
    img {
      display: none;
    }
    width: ${(props) => props.imageBlockWidth};
  }
`;
/**
 * Full width block with a side image this block will take 100% of parent component this component can also be inside a Block component
 */
export default function SideImageBlock(props) {
  return (
    <Container
      className={props.className}
      imageBlockPosition={props.imageBlockPosition}
      style={props.style}
    >
      <ImageBlock
        className={props.className ? props.className + "__image_container" : ""}
        imageBlockWidth={props.imageBlockWidth}
        image={props.image}
      >
        <picture>
          <source
            media={`(min-width:${breakpoints.tablet})`}
            srcSet={props.image.desktop}
          />
          <source
            media={`(min-width:${breakpoints.mobile})`}
            srcSet={props.image.tablet}
          />
          <img src={props.image.mobile} loading="lazy" />
        </picture>
      </ImageBlock>
      <ContentBlock
        className={props.className ? props.className + "__content_block" : ""}
        backgroundColor={props.contentBackgroundColor}
        imageBlockWidth={`${100 - parseInt(props.imageBlockWidth)}%`}
      >
        <ContentContainer
          className={
            props.className ? props.className + "__content_container" : ""
          }
          contentMaxWidth={props.contentMaxWidth}
        >
          {props.children}
        </ContentContainer>
      </ContentBlock>
    </Container>
  );
}
SideImageBlock.defaultProps = {
  imageBlockWidth: "50%",
  imageBlockPosition: "left",
  contentMaxWidth: "auto",
  contentBackgroundColor: "transparent",
  className: "",
};

SideImageBlock.propTypes = {
  className: PropTypes.string,

  /**
   * The content
   */
  children: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.array,
    PropTypes.element,
  ]).isRequired,

  /**
   * Block Image Position
   */
  imageBlockPosition: PropTypes.oneOf(["left", "right"]),

  /**
   * Internal content wrapper Max width CSS width
   */
  contentMaxWidth: PropTypes.string,

  /**
   * Image Block size in persentage
   * */
  imageBlockWidth: PropTypes.string,

  /**
   * HEX CSS Background Color for content block
   */
  contentBackgroundColor: PropTypes.string,

  /**
   * Responsive background image this can be differen aspect ratios per screen sizes
   */
  image: PropTypes.shape({
    mobile: PropTypes.string,
    tablet: PropTypes.string,
    desktop: PropTypes.string,
  }).isRequired,

  /**
  React CSS format to override rules has required by design
  */
  style: PropTypes.object,
};
