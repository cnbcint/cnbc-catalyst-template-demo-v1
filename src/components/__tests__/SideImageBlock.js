import React from "react";
import SideImageBlock from "../SideImageBlock";
import Text from "../Text";
import { render } from "@testing-library/react";
import "jest-styled-components";
import { breakpoints } from "../../helpers/Layout";

const Component = () => (
  <SideImageBlock
    imageBlockPosition="right"
    imageBlockWidth="70%"
    contentMaxWidth="400px"
    contentBackgroundColor="#ccc"
    style={{ marginBottom: 0 }}
    className="test"
    image={{
      desktop:
        "http://localhost:8080/images/responsive/output/image05-desktop.jpg",
      tablet:
        "http://localhost:8080/images/responsive/output/image05-tablet.jpg",
      mobile:
        "http://localhost:8080/images/responsive/output/image05-mobile.jpg",
    }}
  >
    <Text>Testing content</Text>
  </SideImageBlock>
);

beforeAll(() => {
  window.IntersectionObserver = class {
    observe() {
      return null;
    }
    disconnect() {}
  };
});

describe("SideImageBlock", () => {
  it("Renders correctly", () => {
    const { container } = render(<Component />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it("Renders an img with default mobile source", () => {
    const { container } = render(<Component />);
    expect(container.querySelector("img").src).toMatch(
      "http://localhost:8080/images/responsive/output/image05-mobile.jpg"
    );
  });

  it("Add sources for desktop and tablet", () => {
    const { container } = render(<Component />);
    expect(container.querySelectorAll("source")[0].srcset).toMatch(
      "http://localhost:8080/images/responsive/output/image05-desktop.jpg"
    );
    expect(container.querySelectorAll("source")[1].srcset).toMatch(
      "http://localhost:8080/images/responsive/output/image05-tablet.jpg"
    );
  });

  it("Sources has the correct media queries for tablet and desktop", () => {
    const { container } = render(<Component />);
    expect(container.querySelectorAll("source")[0].media).toMatch(
      "(min-width:1280px)"
    );
    expect(container.querySelectorAll("source")[1].media).toMatch(
      "(min-width:768px)"
    );
  });

  it("Renders the correct content", () => {
    const { container } = render(<Component />);
    const target = container.getElementsByClassName(
      "test__content_container"
    )[0];
    expect(target.firstChild.textContent).toMatch("Testing content");
  });

  it("Sets the correct background color", () => {
    const { container } = render(<Component />);
    const target = container.getElementsByClassName("test__content_block")[0];
    expect(target).toHaveStyleRule("background-color", "#ccc");
  });

  it("Sets the correct image width", () => {
    const { container } = render(<Component />);
    const target = container.getElementsByClassName("test__image_container")[0];
    expect(target).toHaveStyleRule("width", "70%", {
      media: `(min-width:${breakpoints.mobile})`,
    });
  });

  it("Sets the correct alignment", () => {
    const { container } = render(<Component />);
    const target = container.getElementsByClassName("test")[0];
    expect(target).toHaveStyleRule("flex-direction", "row-reverse", {
      media: `(min-width:${breakpoints.mobile})`,
    });
  });

  it("Passes the class name", () => {
    const { container } = render(<Component />);
    const target = container.getElementsByClassName("test")[0];
    expect(target.classList.contains("test")).toBe(true);
  });
});

afterAll(() => {
  window.IntersectionObserver = null;
});
