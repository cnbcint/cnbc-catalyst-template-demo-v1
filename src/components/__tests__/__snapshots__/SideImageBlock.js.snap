// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideImageBlock Renders correctly 1`] = `
.c0 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  position: relative;
  box-sizing: border-box;
  margin-top: 3rem;
  margin-bottom: 3rem;
}

.c2 {
  width: 100%;
  padding: 1rem;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #ccc;
}

.c1 {
  background: black;
  background-image: url(http://localhost:8080/images/responsive/output/image05-mobile.jpg);
  background-position: center;
  background-size: cover;
}

.c1 img {
  margin: 0;
}

.c4 {
  font-size: 1.2rem;
  line-height: 125%;
}

@media (min-width:768px) {
  .c0 {
    -webkit-flex-direction: row-reverse;
    -ms-flex-direction: row-reverse;
    flex-direction: row-reverse;
    -webkit-align-items: stretch;
    -webkit-box-align: stretch;
    -ms-flex-align: stretch;
    align-items: stretch;
  }
}

@media (min-width:768px) {
  .c0 {
    margin-top: 5rem;
    margin-bottom: 5rem;
  }
}

@media (min-width:768px) {
  .c2 {
    padding: 3rem;
    width: 30%;
  }
}

@media (min-width:768px) {
  .c3 {
    max-width: 400px;
  }
}

@media (min-width:768px) {
  .c1 {
    width: 70%;
  }

  .c1 img {
    display: none;
  }
}

<div
  class="c0 test"
  style="margin-bottom: 0px;"
>
  <div
    class="c1 test__image_container"
  >
    <picture>
      <source
        media="(min-width:1280px)"
        srcset="http://localhost:8080/images/responsive/output/image05-desktop.jpg"
      />
      <source
        media="(min-width:768px)"
        srcset="http://localhost:8080/images/responsive/output/image05-tablet.jpg"
      />
      <img
        loading="lazy"
        src="http://localhost:8080/images/responsive/output/image05-mobile.jpg"
      />
    </picture>
  </div>
  <div
    class="c2 test__content_block"
  >
    <div
      class="c3 test__content_container"
    >
      <p
        class="c4"
      >
        Testing content
      </p>
    </div>
  </div>
</div>
`;
